keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,144 @@
+Conditions are central enough to the API to make a common golang type with a fixed schema.
+The schema can be a strong recommendation to all API authors.
+
+## Motivation
+
+Allow general consumers to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+The pattern is well-established and we have a good sense of the schema we now want.
+
+### Goals
+
+ 1. For all new APIs, have a common type for `.status.conditions`.
+ 2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+ 3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+### Non-Goals
+
+ 1. Update all existing APIs to make use of the new condition type.
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// Last time the condition transitioned from one status to another.
+    // This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+	// The reason for the condition's last transition in CamelCase.
+    // The specific API may choose whether or not this field is considered a guaranteed API.
+	// +required
+	Reason string `json:"reason" protobuf:"bytes,4,opt,name=reason"`
+	// A human readable message indicating details about the transition.
+    // This field is never considered a guaranteed API and may be empty/missing.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+}
+```
+
+This is not strictly compatible with any of our existing conditions because of either proto ordinals,
+required vs optional, or omitEmpty or not.
+However, it encapsulates the best of what we've learned and will allow new APIs to have a unified type.
+
+### Noteworthy choices
+ 1. `lastTransitionTime` is required.
+    Some current implementations allow this to be missing, but this makes it difficult for consumers.
+    By requiring it, the actor setting the field can set it to the best possible value instead of having clients try to guess.
+ 2. `reason` is required.
+    The actor setting the value should always describe why the condition is the way it is, even if that value is, unknown unknowns.
+    No other actor has the information to make a better choice.
+ 3. `lastHeartbeatTime` is removed.
+    This field caused excessive write loads as we scaled.
```

Comment:
Also, other resource types named this field differently, such as lastProbeTime. I agree that the cost has outweighed the benefit, though.